### PR TITLE
[14285] Add sanitizer workflow

### PR DESCRIPTION
**Pull request by MiguelBarro**
Signed-off-by: Miguel Barro <miguelbarro@eprosima.com>

## Description

The sanitizer fixes go into the intermediate branch `features/tsan/fixes`. The github action must run whenever a *push* or *pull request* is performed against this intermediate branch.

Note that once the  `features/tsan/fixes` branch is merge into master the workflow must be modified to reference `master` changes as trigger.

The action will run on the Ubuntu host using the *apt* provided g++ (thread sanitizer is not yet available on other platforms).

The action's workflow:
- build from sources Fast-DDS and all the dependencies setting the sanitizer flags. 
- run the testing specifying a per-test sanitizer log.
- parsing the result log in order to: identify the issues, sort them by frequency, and filter out duplicates.
- generate a spreadsheet friendly output and a summary of the issues logs (one single report per issue)

A successful workflow run is available [here](https://github.com/MiguelBarro/Fast-DDS/actions/runs/2086601535).

## Contributor Checklist
- [x] Commit messages follow the project guidelines. 
- [x] The code follows the style guidelines of this project. 
- [ ] NA Tests that thoroughly check the new feature have been added/Regression tests checking the bug and its fix have been added. 
- [ ] NA Any new/modified methods have been properly documented using Doxygen. 
- [ ] NA Fast DDS test suite has been run locally. 
- [x] Changes are ABI compatible. 
- [x] Changes are API compatible. 
- [ ] NA Documentation builds and tests pass locally. 
- [ ] NA New feature has been added to the `versions.md` file (if applicable).
- [ ] NA New feature has been documented/Current behavior is correctly described in the documentation. 

## Reviewer Checklist
- [ ] Check contributor checklist is correct.
- [ ] Check CI results: changes do not issue any warning.
- [ ] Check CI results: failing tests are unrelated with the changes.

### CODE REVIEW

**Comment by MiguelCompany** [comment]
@richiprosima Please test this

File: .github/workflows/thread-sanitizer.yaml [hunk 1]
```diff
@@ -0,0 +1,163 @@
+name: Thread Sanitizer analysis
+on:
+  push:
```
Comment (rsanchez15) [inline .github/workflows/thread-sanitizer.yaml:3]: Set this so we can launch it manually at any time
```suggestion
  workflow_dispatch:
  push:
```

File: .github/workflows/thread-sanitizer.yaml [hunk 2]
```diff
@@ -0,0 +1,163 @@
...
+          # Export test mappings
+          & {$sd; $sr} | Sort-Object { [int]$_.fuzzhash} | % { $fuzzy=$_.fuzzhash;$_.tests |
+            select @{l="fuzzhash";e={$fuzzy}}, @{l="test";e={$_}}} |
+            Export-Csv (Join-Path $exports issue_test_map.csv)
+          # Keep a file per issue
+          $dir = New-Item -ItemType Directory -Path (Join-Path $exports reports)
+          & {$sd; $sr} | % { $_.échantillon | Out-File (Join-Path $dir "$($_.fuzzhash).tsan") }
+          # Zip an artifact
+          popd
+          gci $exports | Compress-Archive -DestinationPath ./ws/artifact.zip
```
Comment (rsanchez15) [inline .github/workflows/thread-sanitizer.yaml:157]:
```suggestion
          gci $exports | Compress-Archive -DestinationPath ./ws/tsan_reports.zip 
```

File: .github/workflows/thread-sanitizer.yaml [hunk 3]
```diff
@@ -0,0 +1,163 @@
...
+          & {$sd; $sr} | % { $_.échantillon | Out-File (Join-Path $dir "$($_.fuzzhash).tsan") }
+          # Zip an artifact
+          popd
+          gci $exports | Compress-Archive -DestinationPath ./ws/artifact.zip 
+
+      - name: Archive sanitation results
+        uses: actions/upload-artifact@v3
+        with:
+            name: sanitation-report
+            path: /home/runner/work/ws/artifact.zip
```
Comment (rsanchez15) [inline .github/workflows/thread-sanitizer.yaml:163]:
```suggestion
            path: /home/runner/work/ws/tsan_reports.zip
```

**Review by rsanchez15 — CHANGES_REQUESTED** [review]

**Comment by rsanchez15** [comment]
I ran this action on a Fast DDS fork repository and [here](https://github.com/rsanchez15/Fast-DDS/actions/runs/2102497837) are the results.

Please check also that almost all Fast DDS tests fail.

**Comment by MiguelBarro** [comment]
As it is the workflow never fails. It runs all the tests and files the reports. In the future, when we manage to clear all the reports, we may enforce failure if new reports appear. Currently doesn't make much sense.

**Review by rsanchez15 — APPROVED** [review]

**Pull request merged**